parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,546 @@
+      this.key = key;
+      this.value = value;
+    }
+  }
+
+  /**
+   * Returns the ObjectField at the `index` slot. Return null if `index` is out of the bound of
+   * `[0, objectSize())`. `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get
+   * @return the Objectfield at the `index` slot, or null if `index` is out of bounds
+   */
+  public ObjectField getFieldAtIndex(int index) {
+    return VariantUtil.handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        return null;
+      }
+      return getFieldAtIndex(index, value, metadata, idSize, offsetSize, idStart, offsetStart, dataStart);
+    });
+  }
+
+  private static ObjectField getFieldAtIndex(
+      int index,
+      byte[] value,
+      byte[] metadata,
+      int idSize,
+      int offsetSize,
+      int idStart,
+      int offsetStart,
+      int dataStart) {
+    int id = VariantUtil.readUnsigned(value, idStart + idSize * index, idSize);
+    int offset = VariantUtil.readUnsigned(value, offsetStart + offsetSize * index, offsetSize);
+    String key = VariantUtil.getMetadataKey(metadata, id);
+    Variant v = new Variant(value, metadata, dataStart + offset);
+    return new ObjectField(key, v);
+  }
+
+  /**
+   * Returns the dictionary ID for the object field at the `index` slot.
+   * `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get the dictionary ID for
+   * @return the dictionary ID for the object field at the `index` slot
+   * @throws MalformedVariantException if `index` is out of bounds
+   */
+  public int getDictionaryIdAtIndex(int index) {
+    return VariantUtil.handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        throw VariantUtil.malformedVariant();
+      }
+      return VariantUtil.readUnsigned(value, idStart + idSize * index, idSize);
+    });
+  }
+
+  /**
+   * @return the number of array elements. `getType()` must be `Type.ARRAY`.
+   */
+  public int arraySize() {
+    return VariantUtil.handleArray(value, pos, (size, offsetSize, offsetStart, dataStart) -> size);
+  }
+
+  /**
+   * Returns the array element Variant value at the `index` slot. Returns null if `index` is
+   * out of the bound of `[0, arraySize())`. `getType()` must be `Type.ARRAY`.
+   * @param index the index of the array element to get
+   * @return the array element Variant at the `index` slot, or null if `index` is out of bounds
+   */
+  public Variant getElementAtIndex(int index) {
+    return VariantUtil.handleArray(value, pos, (size, offsetSize, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        return null;
+      }
+      return getElementAtIndex(index, value, metadata, offsetSize, offsetStart, dataStart);
+    });
+  }
+
+  private static Variant getElementAtIndex(
```

Comment:
This is only used by `getElementAtIndex(int)` and in the JSON conversion code. The method above, `getElementAtIndex(int)` isn't used at all, nor is `arraySize`. Does that mean this is not tested?